src/compiler/crystal/semantic/restrictions.cr

```diff
@@ -504,7 +504,29 @@ module Crystal
       # Special case: consider `Union(X, Y, ...)` the same as `X | Y | ...`
       generic_type = get_generic_type(other, context)
       if generic_type.is_a?(GenericUnionType)
-        return restrict(Union.new(other.type_vars), context)
+        types = [] of Type
+
+        other.type_vars.each do |type_var|
+          if type_var.is_a?(Splat)
+            splat_type = context.defining_type.lookup_type?(type_var.exp)
+            return nil unless splat_type
+            unless splat_type.is_a?(TupleInstanceType)
+              type_var.raise "argument to splat must be a tuple type, not #{splat_type}"
+            end
+
+            splat_type.tuple_types.each do |tuple_type|
+              if type = restrict(tuple_type, context)
```

Comment:
What happens when the restriction doesn't match? Shouldn't the entire method return `nil`?

Reply:
If any restriction inside a splat doesn't match or even if the entire splat's element types don't match, there may still be other type vars in `Union(...)`:

```crystal
class Foo(*T)
  def self.f(x : Union(*T, Bool)); end
end

Foo(Int32, String).f(true) # okay even if !(Bool < Union(*T))
```

Compare this to the case for a regular union, a few dozen lines above:

```crystal
def restrict(other : Union, context)
  types = other.types.compact_map do |ident|
    restrict(ident, context).as(Type?)
  end
  types.size > 0 ? program.type_merge_union_of(types) : nil
end
```

So dropping failed restrictions here is fine.

Reply:
But what if one type fails to match and the next one in the tuple succeeds? Will that match?

Reply:
Yes, I just didn't write out the `flat_map` + inner `compact_map` to save array allocations. There is probably room for a few additional specs.

Reply:
I totally missed this was for `Union(...)` (I reviewed this without enough time so I just skimmed through the code). Now I understand this better. Thanks for bearing with me!